File: Neetcode Questions/Trees/validate-binary-search-tree.cpp

// https://leetcode.com/problems/validate-binary-search-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBSTHelper(TreeNode* root, TreeNode*minNode, TreeNode*maxNode){
        if(!root) return true;
        if((minNode && root->val <= minNode->val) || (maxNode && root->val >= maxNode->val)) return false;
        return isValidBSTHelper(root->left, minNode, root) && isValidBSTHelper(root->right, root, maxNode); 
    }

    bool isValidBST(TreeNode* root) {
        return isValidBSTHelper(root, NULL, NULL);
    }
};

File: Neetcode-150-js/Trees/validate-binary-search-tree.js

// https://leetcode.com/problems/validate-binary-search-tree/description/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    if(!root) return true;

    var solve = (root, minNode, maxNode) => {
        if(!root) return true;
        if(root.val <= minNode || root.val >= maxNode) return false;
        return solve(root.left, minNode, root.val) && solve(root.right, root.val, maxNode);
    }

    return solve(root, Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER);
};